src/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java

```diff
@@ -491,7 +491,7 @@ private boolean routeToRemoteDomain(JID jid, Packet packet,
 		    boolean retry = false;
 		    // If we're here, it means we have no functional route. Sort it out.
 		    final String domain = jid.getDomain();
-		    synchronized (domain.intern()) { // Only create one route at a time.
+		    synchronized (("AddRoute::" + domain).intern()) { // Only create one route at a time.

```

Comment:
I know it's beyond the scope of this PR, but it's worth noting that S2S connections are shared by all nodes in a cluster configuration. This should really be constrained using a shared cache lock rather than a local synchronized block. There has been some recent chatter related to this in the community forum.